generic-ec/src/multiscalar.rs

```diff
@@ -0,0 +1,406 @@
+//! Multiscalar multiplication
+//!
+//! Let $s_{1, \dots, n}$ and $P_{1, \dots, n}$ be lists of scalars and points
+//! respectively. Multiscalar multiplication is computing point $Q$ such that:
+//!
+//! $$Q = s_1 P_1 + \dots + s_n P_n$$
+//!
+//! This module provides various algorithms for computing multiscalar multiplication
+//! efficiently
+//!
+//! ## Motivation
+//! Computing the sum naively, i.e. calculating each $s_i P_i$ separately and
+//! $\sum$-ing them, is inefficient even for small $n$.
+//!
+//! Recall an algorithm for computing $Q = s P$:
+//! 1. Let $Q \gets \O, A \gets P$
+//! 2. While $s \ne 0$ do the following:
+//!    1. If $s$ is odd, set $Q \gets Q + A$
+//!    2. Set $s \gets s \gg 1$ (bitwise shift to left by 1 bit)
+//!    3. If $s \ne 0$, set $A \gets A + A$
+//! 3. Return $Q$
+//!
+//! This algorithm does 1 point doubling per each bit in $s$ and one addition
+//! per each bit in $s$ that's set to `1`. For a random scalar, we can expect
+//! that, on average, half of its bits are set to `1`. Thus, cost of the
+//! multiplication is:
+//!
+//! $$\text{cost} = \log_2 s \cdot D + \frac{1}{2} \log_2 s \cdot A$$
+//!
+//! i.e. it does $\log_2 s$ doublings and $\frac{1}{2} \log_2 s$ additions. Naive
+//! multiscalar multiplication algorithm just computes $s_i P_i$ individually
+//! and then sums them. Total cost is:
+//!
+//! $$\text{cost} = n A + n (\log_2 s \cdot D + \frac{1}{2} \log_2 s \cdot A)$$
+//!
+//! Typically, we're working with scalars of 256 bits size. You can easily calculate
+//! that, for instance, even for $n = 3$, the total cost is $768 D + 387 A$. However,
+//! for the same $n$, Straus algorithm has $\text{cost} = 300 D + 192 A$ which
+//! is already significantly faster than naive algorithm.
+//!
+//! You can see the perfromance comparison for different algorithms on secp256k1 curve
+//! on the plot below:
+//!
+//! <img src="https://raw.githubusercontent.com/dfns/generic-ec/m/perf/multiscalar/secp256k1.svg"/>
```

Comment:
Probably better to use explicit commit in the link

Reply:
But then I will forget to update it if it gets changed

Reply:
<img width="768" alt="Screenshot 2024-04-05 at 12 09 00" src="https://github.com/dfns/generic-ec/assets/6142210/f68eed7b-4b46-4d86-b5ea-9622b06c17fd">

I mean

Reply:
I want to say that even if you update it, do you really want the old docs to use the new graph? Also I'm pretty sure in the future we will move those pictures around and the docs will break.

Also, here's one alternative solution: https://users.rust-lang.org/t/include-images-in-rustdoc-output/3487/4

Reply:
Another solution: since it's an svg, you can just insert it literally or as base64 into the docstring

Reply:
I changed it to include svg into the docs